lightning/src/ln/channel.rs

```diff
@@ -2222,6 +2222,102 @@ impl<SP: Deref> InitialRemoteCommitmentReceiver<SP> for FundedChannel<SP> where
 }
 
 impl<SP: Deref> PendingV2Channel<SP> where SP::Target: SignerProvider {
+	/// Prepare and start interactive transaction negotiation.
+	/// `change_destination_opt` - Optional destination for optional change; if None,
+	///   default destination address is used.
+	/// If error occurs, it is caused by our side, not the counterparty.
+	#[allow(dead_code)] // TODO(dual_funding): Remove once contribution to V2 channels is enabled
+	fn begin_interactive_funding_tx_construction<ES: Deref>(
+		&mut self, signer_provider: &SP, entropy_source: &ES, holder_node_id: PublicKey,
+		change_destination_opt: Option<ScriptBuf>,
+		prev_funding_input: Option<(TxIn, TransactionU16LenLimited)>,
+	) -> Result<Option<InteractiveTxMessageSend>, AbortReason>
+	where ES::Target: EntropySource
+	{
+		debug_assert!(self.interactive_tx_constructor.is_none());
+
+		let mut funding_inputs = Vec::new();
+		mem::swap(&mut self.dual_funding_context.our_funding_inputs, &mut funding_inputs);
+
+		if let Some(prev_funding_input) = prev_funding_input {
+			funding_inputs.push(prev_funding_input);
+		}
+
+		let funding_inputs_prev_outputs = DualFundingChannelContext::txouts_from_input_prev_txs(&funding_inputs)?;
```

Comment:
We could use an `Iterator` and `chain` these instead. The error checking in `txouts_from_input_prev_txs` could move to `calculate_change_output_value` if that took an `Iterator` instead of a `Vec`. That would avoid some heap allocations.

Reply:
`funding_inputs` is passed below into `InteractiveTxConstructorArgs` (which takes ownership of the `Vec`), so your proposed optimization is not possible for that reason.

Reply:
Hmm? `calculate_change_output_value` takes the `Vec` by reference. That's why I was saying to take an `Iterator` there.

Reply:
Sorry, forgot to mark this unresolved.

Reply:
The vector, optionally extended with the optional extra input, is needed to be built, since it is stored in the `InteractiveTxConstructorArgs`. If the vector has to be built anyways, there is not much difference between passing a reference to the vector, or an iterator from the vector, is it? It it was possible to use a chained iterator, that would be different.

Reply:
This is moot now, I have inlined the prev tx check.